core/src/main/kotlin/xtdb/expression/map/RelationMap.kt

```diff
@@ -163,38 +177,31 @@ class RelationMap(
                 andIBO(acc, comp)
             }) ?: andIBO()
         }
-        
+
         val hasher = createHasher(probeKeyCols)
-        
+
         return object : RelationMapProber {
+            // TODO one could very likely do a similar thing to a merge sort phase with the buildHashTrie and a probeHashTrie
             override fun indexOf(inIdx: Int, removeOnMatch: Boolean): Int{
                 val hashCode = hasher.hashCode(inIdx)
-                val hashBitmap = hashToBitmap.get(hashCode)
-                
-                return hashBitmap.findInHashBitmap(comparator, inIdx, removeOnMatch)
+                return buildHashTrie.findValue(hashCode, { testIdx -> comparator.applyAsInt(inIdx, testIdx) }, removeOnMatch)
             }
-            
+
             override fun forEachMatch(inIdx: Int, c: IntConsumer) {
                 val hashCode = hasher.hashCode(inIdx)
-                val hashBitmap = hashToBitmap.get(hashCode)
-                
-                hashBitmap?.let { bitmap ->
-                    val iterator = bitmap.intIterator
-                    while (iterator.hasNext()) {
-                        val outIdx = iterator.next()
-                        if (comparator.applyAsInt(inIdx.toInt(), outIdx) == 1) {
-                            c.accept(outIdx)
-                        }
+                for (outIdx in buildHashTrie.findCandidates(hashCode)) {
+                    if (comparator.applyAsInt(inIdx, outIdx) == 1) {
```

Comment:
`comparator` naming is misleading here, especially with `applyAsInt` - I spent most of the PR review assuming this was a `java.util.Comparator`, or at least a function that returned positive for >, 0 for =, negative for < 😅 

it seems you're effectively providing an IntPredicate instead?